lib/semantics/check-do-concurrent.cc

```diff
@@ -0,0 +1,205 @@
+                                         " in DO CONCURRENT"_err_en_US});
+      }
+    }
+  }
+
+  // C1138: extended ranges in DOs should be errors, not warnings
+
+  // 11.1.7.5
+  void Post(const parser::IoControlSpec &ioControlSpec) {
+    if (auto *charExpr =
+            std::get_if<parser::IoControlSpec::CharExpr>(&ioControlSpec.u)) {
+      if (std::get<parser::IoControlSpec::CharExpr::Kind>(charExpr->t) ==
+          parser::IoControlSpec::CharExpr::Kind::Advance) {
+        messages_.Say(charBlock_,
+            parser::MessageFormattedText{
+                "ADVANCE specifier not allowed in DO CONCURRENT"_err_en_US});
+      }
+    }
+  }
+
+private:
+  bool ObjectIsCoarray() { return false; } // placeholder
+  bool EndTDeallocatesCoarray() { return false; } // placeholder
+
+  parser::CharBlock charBlock_;
+  parser::Messages &messages_;
+};
+
+class DoConcurrentCollection {
+public:
+  DoConcurrentCollection(parser::Messages &messages) : messages_{messages} {}
+  template<typename T> constexpr bool Pre(const T &) { return true; }
+  template<typename T> constexpr void Post(const T &) {}
+  bool Pre(const parser::ExecutionPart &executionPart) {
+    const auto &cend{executionPart.v.cend()};
+    for (auto iter{executionPart.v.cbegin()}; iter != cend; ++iter) {
+      CheckDoConcurrent(iter);
+      currentIter_ = iter;
+    }
+    return true;
+  }
+  template<typename T> void Post(const parser::Statement<T> &statement) {
+    if (!labels_.empty() && statement.label.has_value() &&
+        labels_.back() == *statement.label) {
+      CheckConstraints(++labelDoIters_.back(), currentIter_);
+      labels_.pop_back();
+      labelDoIters_.pop_back();
+    }
+  }
+
+private:
+  void CheckConstraints(const parser::Block::const_iterator &begin,
+      const parser::Block::const_iterator &end) {
+    DoConcurrentEnforcement doConcurrentEnforcement{messages_};
+    for (auto iter = begin; iter != end; ++iter) {
+      Walk(*iter, doConcurrentEnforcement);
+    }
+  }
+  void CheckDoConcurrent(
+      const std::list<parser::ExecutionPartConstruct>::const_iterator &iter) {
+    const parser::ExecutionPartConstruct &executionPartConstruct{*iter};
+    if (auto *executableConstruct = std::get_if<parser::ExecutableConstruct>(
+            &executionPartConstruct.u)) {
+      if (auto *doConstruct =
+              std::get_if<common::Indirection<parser::DoConstruct>>(
+                  &executableConstruct->u)) {
+        if (std::get<std::optional<parser::LoopControl>>(
+                std::get<parser::Statement<parser::NonLabelDoStmt>>(
+                    (*doConstruct)->t)
+                    .statement.t)
+                .has_value()) {
+          CheckConstraints(std::get<parser::Block>((*doConstruct)->t).cbegin(),
+              std::get<parser::Block>((*doConstruct)->t).cend());
+        }
+      } else if (auto *labelDoLoop = std::get_if<parser::Statement<
```

Comment:
Why would there still be label DO statements in the program?  Aren't they canonicalized into block DO constructs by now?

Reply:
I implemented this pass first. No other reason.